feat: add new event type to WormholeConnectEventCore

bradleyplaydon opened the pull request:
**Introduces breaking changes**:
- Updated defaultInput prop from 'fromChain, toChain, fromToken, toToken' to 'source' and 'destination'

**Updates**
- Added a new event type to WormholeConnectEventCore to support remembering the last chain and token the user selected so if they refresh it persists
- Fix condition which inside validateDefaults as it no longer stands because we support same chain swaps

bradleyplaydon committed:
feat: add new event type to WormholeConnectEventCore

bradleyplaydon committed:
refactor: added telementry utils

bradleyplaydon committed:
build: fix build

kev1n-peters commented on src/telemetry/utils.ts:
```diff
@@ -0,0 +1,35 @@
+export const handleSelectChain = (
+  chain: Chain,
+  isSource: boolean,
+  config: InternalConfig<'Mainnet' | 'Testnet' | 'Devnet'>,
```
Can these just be `InternalConfig<Network>`?

priyajeet replied:
Also can we just import this? instead of passing in

bradleyplaydon replied:
> Also can we just import this? 

Wouldn't this be counter intuitive since it's already imported in the component and being used for other things not just telemetry like the useTokenList hook

bradleyplaydon replied:
> > Also can we just import this?
> 
> Wouldn't this be counter intuitive since it's already imported in the component and being used for other things not just telemetry like the useTokenList hook

Unless I do a dynamic import based upon if the param is passed and make it optional?

bradleyplaydon replied:
For now leaving as it, think if we run into multiple cases where configs having to be imported into the component that imports these functions we can import it here

priyajeet commented on src/views/v2/Bridge/AssetPicker/index.tsx:
```diff
@@ -26,6 +26,7 @@ import TokenList from './TokenList';
 import AssetBadge from 'components/AssetBadge';
 import type { Token } from 'config/tokens';
 import { useTokenList } from 'hooks/useTokenList';
+import { handleSelectChain, handleSelectToken } from 'telemetry/utils';
```
Need to update the v3 (views/v3) folder also

priyajeet replied:
I won't even bother changing V2 at this point if it makes life simpler for new features

priyajeet commented on src/telemetry/utils.ts:
```diff
@@ -0,0 +1,35 @@
+  });
+};
+
+export const handleSelectToken = (
```
Should we also rename the exported functions to indicate they are telemetry related? Since these words are common and can also be used in other places with same meaning. 

Maybe something along the lines of `handleTelemetryOnChainSelect` and `handleTelemetryOnTokenSelect`. Something like this will likely avoid casting to a new name if there is a conflict + person reading them will not need to see where its imported from to understand usecase/functionality

bradleyplaydon replied:
Love this, the more explicit the better imo

bradleyplaydon committed:
refactor: util naming convention and use more specific type

bradleyplaydon committed:
refactor: add telementry function calls to v3 asset picker

bradleyplaydon committed:
feat: add helper type to help narrow the value type by action

bradleyplaydon committed:
refactor: remove 'const' from enum to fix importing

bradleyplaydon committed:
refactor: make DefaultInputs type clearer with chain being a required field

Before you could pass to the config just a token in defaultInputs object but everywhere
defaultInputs is referenced chain is required and you can't just pass a token

bradleyplaydon committed:
feat: add handleTelementryOnChainSelect to selecting token incase the user doesn't select chain

bradleyplaydon committed:
feat: remove config param from telemetry calls

priyajeet commented on src/telemetry/utils.ts:
```diff
@@ -0,0 +1,33 @@
+import config from 'config';
+import type { Chain, Token } from 'exports';
```
Lets fix these imports. Should come from the actual file next to this file using relative import and not exports module. Chain/Token should come from wormhole sdk or wherever?

bradleyplaydon committed:
refactor: update naming convention for default inputs

priyajeet commented on src/config/utils.ts:
```diff
@@ -135,72 +135,76 @@ export const validateDefaults = (
-  if (defaults.fromChain && defaults.toChain) {
-    if (defaults.fromChain === defaults.toChain) {
+  if (defaults.source?.chain && defaults.destination?.chain) {
+    if (defaults.source.chain === defaults.destination.chain) {
```
change check to token check, not chain check. Chains can be same

bradleyplaydon committed:
fix: fix condition not being relevant because of same chain swaps
